---

title: "exploratory gene differential expression between patients with squamous cell carcinoma vs cancer free indiviuals"
author: "sara yusuf , Salma yahia"
date: "2023-08-25"
output: html_document
  
    
---



    Disclosure

This project is done by "Sara yusuf" and "Salma yahia" as a graduation project.
course         : data analysis and visulaisation with R "DAV-R" 
supervised by  : "Dr-Fadhl Alakwaa" and "Dr-Mohammed Hamed"
year           : 2023



                               

    Dataset Description

We have done an RNA-seq Differential gene expression and enrichment analysis using Lung squamous cell carcinoma data set.
an RNA quantification was done by RNAseq (HiSeq) experiment to quantify 21792 genes from 108 samples with a confirmed diagnosis with LSCC and 94 healthy individuals.

    dataset source

"https://linkedomics.org/data_download/CPTAC-LSCC/"


    Analysis workflow


1. Data importing, exploration, wrangling and quality control.
2. performing PCA 
3. samples clustering  
4. DEGs analysis
5. gene enrichment  
6. statistical analysis on the top differentiated gene

__________________________________________________________________________________________________________________________

  
calling the libraries and Dataset uploading 

```{r}
library(dplyr )
library(ggplot2)
library(tidyverse)
library(impute)
library(vegan)
library(cluster)
library(factoextra)
library(gridExtra)
library(dplyr)
library(FactoMineR)
library(useful)
library(limma)
library(enrichR)
library(ggpubr)
library(EnhancedVolcano)

```


```{r}
mytheme=theme(
        plot.title = element_text(hjust = 0.5,size=30,face = "bold"),
        axis.text.x = element_text( size = 15, angle = 45, hjust = .5, vjust = 0.5, face = "plain"),
        axis.text.y = element_text( size = 15, angle = 0, hjust = 1, vjust = 0, face = "plain"),
        axis.title.x = element_text( size = 20, angle = 0, hjust = .5, vjust = 0, face = "bold"),
        axis.title.y = element_text( size = 20, angle = 90, hjust = .5, vjust = .5, face = "bold"),
        #legend.title=element_text(size=20),
        legend.title=element_blank(), # remove legend title name
        legend.text = element_text(size=15,face="plain"),
        strip.text = element_text(size = 15,face="plain") ,
        legend.position="right",

        # for transparent background
        panel.background = element_rect(fill = "transparent"), # bg of the panel
        plot.background = element_rect(fill = "transparent", color = NA), # bg of the plotge
        panel.grid.major = element_blank(), # get rid of major grid
        panel.grid.minor = element_blank(), # get rid of minor grid
        legend.background = element_rect(fill = "transparent"), # get rid of legend bg
        legend.box.background = element_rect(fill = "transparent"), # get rid of legend panel bg
        axis.line = element_line(colour = "black") # adding a black line for x and y axis
)

```


```{r}
rna.normal = read.csv("C:/Users/saraa/Desktop/delss/lscc/rna normal.txt",row.names =1, sep = "\t")
rna.tumor = read.csv("C:/Users/saraa/Desktop/delss/lscc/rna tumor.txt",row.names =1, sep = "\t")
normal.clinical = read.csv("C:/Users/saraa/Desktop/delss/lscc/normal.txt" ,row.names =1, sep = "\t")
tumor.clinical = read.csv("C:/Users/saraa/Desktop/delss/lscc/tumor.txt",row.names =1, sep = "\t")
```



```{r}
rna.normal = t(rna.normal)
rna.tumor = t(rna.tumor)
```





```{r}
common_col = intersect(colnames(rna.tumor), colnames(rna.normal))

rna_mat =rbind(subset(rna.tumor, select = common_col ),
  subset(rna.normal, select = common_col))

```



```{r}
class(rna_mat)
```


```{r}
dim(rna_mat)
```
```{r}
rna_mat = rna_mat%>%
  data.frame%>%
  rownames_to_column("id")

```


Data joining to create a comprehensive data frame


```{r}
normal.clinical = normal.clinical %>%
data.frame%>%
rownames_to_column("id")

```


```{r}
tumor.clinical = tumor.clinical %>%
data.frame%>%
rownames_to_column("id")
```

```{r}
normal.data= rna.normal %>%
data.frame %>%
rownames_to_column('id') %>%
inner_join(.,normal.clinical ,by='id')
```

```{r}
dim(normal.data)  

```


```{r}
tumor.data= rna.tumor %>%
data.frame %>%
rownames_to_column('id') %>%
inner_join(.,tumor.clinical ,by='id')
```

```{r}
dim(tumor.data)

```


```{r}
normal.data = normal.data%>%
mutate(label = "normal")

```


```{r}
normal.data = normal.data%>%
mutate(Stage = "no stage")
```


```{r}
dim(normal.data)
```


```{r}
tumor.data = tumor.data%>%
mutate(label = "tumor")
```

```{r}
dim(tumor.data)

```
```{r}
length(intersect(colnames(tumor.data), colnames(normal.data)))
```


```{r}
common_cols = intersect(colnames(tumor.data), colnames(normal.data))

final_data =rbind(subset(tumor.data, select = common_cols ),
  subset(normal.data, select = common_cols))
```

converting the categorical variables to numeric

```{r}
final_data =  transform(final_data, Age = as.numeric(Age),
       
                                Pack.Years.Smoked = as.numeric(Pack.Years.Smoked))
```



```{r}
str(final_data$Age)
```

```{r}
 rownames(final_data) %in% rownames(rna_mat)
```

```{r}
all(rownames(final_data) == rownames(rna_mat))
```

```{r}
rna_mat = as.matrix(rna_mat)
```


calculate the variance of each column and remove coloumns wit zero var

```{r}

varCol=apply(rna_mat, 2, var, na.rm = T)
```



```{r}
constCol = (varCol == 0 | is.na(varCol) )
```


```{r}
rna_mat = rna_mat[, !constCol]
```

calculate the rate of missing data and drop columns with a rate more than 50% missing values.

```{r}
missingness_rate = apply(is.na(rna_mat), 2, sum)/nrow(rna_mat) *100
```


```{r}
max(missingness_rate)
```

visualise the missingness rate


```{r}
h=hist(missingness_rate,breaks=10,
main="",
xlab="percentage of missingness")

```
```{r}
good.col=missingness_rate < 50

rna_mat=rna_mat[,good.col]
```


impute missing values with knn  



```{r}
rna_mat_imputed=impute.knn(rna_mat,k=10)$data

class(rna_mat_imputed)

dim(rna_mat_imputed)

sum(is.null(rna_mat_imputed))
```



 normalise the data with log2 transform

```{r}
rna_mat_imputed.logged = log2(rna_mat_imputed + 1)

```

```{r}
par(mfrow=c(1,2))
plot(density(apply(rna_mat_imputed, 2, mean, na.rm = TRUE)),main='befor log2')
plot(density(apply(rna_mat_imputed.logged, 2, mean, na.rm = TRUE)),main='after log2')
```

```{r}
data = as.data.frame(rna_mat_imputed.logged)

```



```{r}
df_pca = prcomp(data)
df_out = as.data.frame(df_pca$x)
ggplot(df_out,aes(x=PC1,y=PC2,color=final_data$label,shape=final_data$label ))+
geom_point()+ggtitle("")+labs(color='')+
  geom_point(size=8,alpha=0.5)
```


as seen the data is normally distributed and no significant difference noticed between before and after log transform and no outliers were detected by PCA



 first clustering using kmeans , define number of centers = 2


```{r}
kmeans2 = kmeans((data), centers = 2, nstart = 25)



```


```{r}
kmeans2$cluster
```



```{r}
fviz_cluster(kmeans2, data = data, ellipse = T,labelsize = 5)
```


```{r}
plot(kmeans2, data=data)
```


```{r}
table(final_data$label, kmeans2$cluster)

```



```{r}
plot(table(final_data$label, kmeans2$cluster))

```

 samples are perfectly clustered comparing with the original labels without overlaping


```{r}
kmeans3 = kmeans(data, centers = 3, nstart = 25)  
kmeans4 = kmeans(data, centers = 4, nstart = 25)  
kmeans5 = kmeans(data, centers = 5, nstart = 25)  
 #Comparing the Plots
plot1 = fviz_cluster(kmeans2, geom = "point", data = data) + ggtitle("k = 2")
plot2 = fviz_cluster(kmeans3, geom = "point", data = data) + ggtitle("k = 3")
plot3 = fviz_cluster(kmeans4, geom = "point", data = data) + ggtitle("k = 4")
plot4 = fviz_cluster(kmeans5, geom = "point", data = data) + ggtitle("k = 5")
grid.arrange(plot1, plot2, plot3, plot4, nrow = 2)
```



  plot the error with number of clusters to predict the optimal clusters by "wss" and  "silhouette" methods


```{r}
set.seed(123)

fviz_nbclust(data, kmeans, method = "wss")
```


```{r}
fviz_nbclust(data, kmeans, method = "silhouette")

```

 depending on the results a conclusion is made that the optimal number of clusters is 2



 the second method of clustering by "Hierarchical Clustering"

```{r}

d = dist(data, method = "euclidean") 
fit = hclust(d, method="average")
plot(fit)
groups = cutree(fit, k=2)

rect.hclust(fit, k=2, border="red")
```




```{r}
class(data)
```


 now proceeding with differential gene expression analysis by limma 

```{r}
type = as.character(final_data$label)

design <- model.matrix(~0+factor(type))

colnames(design) <- levels(factor(type))
```

```{r}
contrast<-makeContrasts(tumor-normal,levels=design)
fit <- lmFit(as.matrix(t(data)), design)
fit2 <- contrasts.fit(fit, contrast)
fit2 <- eBayes(fit2)
```


```{r}
DEGs=topTable(fit2, adjust.method ='fdr', number = 999999999, p.value = 0.05, coef = 1)
dim(DEGs)
head(DEGs)
names(fit)
names(DEGs)
```



```{r}
DEGs%>%
 filter(adj.P.Val < 0.05 & logFC >0 )%>%
  count
  
  
```

there is 8446 genes are up-regulated among cancer cases with adj.p < 0.05


```{r}
EnhancedVolcano(DEGs,
    lab =  rownames(DEGs),
    x   = 'logFC' ,
    y   = 'P.Value',
    title = 'lung cancer vs Healthy',
    pCutoff = 10e-32,
    FCcutoff = .5,
    pointSize = 3.0,
    labSize =2.0)

```






    gene enrichment analysis by enrichR bioconducter package.

 the  databases to search in:
  
"Reactome_2016"
"GO_Biological_Process_2021"
"GO_Cellular_Component_2021"
"GO_Molecular_Function_2021"


```{r}
dbs = c("Reactome_2016","GO_Biological_Process_2021","GO_Cellular_Component_2021","GO_Molecular_Function_2021")

```

gene enrichment for the significantly up regulated genes


```{r}
d1=DEGs %>% filter( adj.P.Val <0.05 & logFC >0) %>% rownames
    
enriched_up = enrichr(d1 , dbs)


```




```{r}

d = plotEnrich(enriched_up[[1]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value",
               title=paste0(length(d1),'_','','_','_',dbs[1]))
print(d)
 

d = plotEnrich(enriched_up[[2]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value",
               title=paste0(length(d1),'_','','_','_',dbs[2]))
 print(d)

d = plotEnrich(enriched_up[[3]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value",
               title=paste0(length(d1),'_','','_','_',dbs[3]))
 print(d)
d = plotEnrich(enriched_up[[4]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value",
               title=paste0(length(d1),'_','','_',dbs[4]))
print(d)

```

gene enrichment for the significantly down regulated genes


```{r}
d2=DEGs %>% filter( adj.P.Val <0.05 & logFC <0) %>% rownames
    
enriched_down = enrichr(d2 , dbs)

```

```{r}
d = plotEnrich(enriched_down[[1]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value",
               title=paste0(length(d1),'_','','_','_',dbs[1]))
print(d)
 

d = plotEnrich(enriched_down[[2]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value",
               title=paste0(length(d1),'_','','_','_',dbs[2]))
 print(d)

d = plotEnrich(enriched_down[[3]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value",
               title=paste0(length(d1),'_','','_','_',dbs[3]))
 print(d)
d = plotEnrich(enriched_down[[4]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value",
               title=paste0(length(d1),'_','','_',dbs[4]))
print(d)
```


 plotting the label against the top differentiated gene "POLQ	" between normal samples vs  samples with lung cancer , then apply  t-test and "wilcox.test" for comparison

```{r}
ggplot(data,aes(final_data$label, POLQ	 ,color=final_data$label))+geom_boxplot()+mytheme+geom_jitter(aes(color=final_data$label))+
stat_summary(fun.y="mean", color="black", geom="point",shape=18, size=5,position=position_dodge(width=0.75))+
stat_compare_means( method = "t.test",size = 5)
```


```{r}
ggplot(data,aes(final_data$label, POLQ	,color=final_data$label))+geom_boxplot()+mytheme+geom_jitter(aes(color=final_data$label))+
stat_summary(fun.y="mean", color="black", geom="point",shape=18, size=5,position=position_dodge(width=0.75))+
stat_compare_means( method = "wilcox.test",size = 5)

```

 it is perfectly clear that there is a significant difference between the two groups


 two way anova test to rule out if gender and age are co factors between  normal samples and samples with lung cancer

```{r}
two.way = aov(POLQ~final_data$label+final_data$Age+final_data$Gender,data=data)
summary(two.way)


```

 There is no role of age or gender in the risk of developing lung cancer

```{r}
ggplot(data,aes(final_data$label, POLQ	,color=final_data$label))+geom_boxplot()+mytheme+geom_jitter(aes(color=final_data$label))+
stat_summary(fun.y="mean", color="black", geom="point",shape=18, size=5,position=position_dodge(width=0.75))+
stat_compare_means( label="p.signif",method = "t.test",size = 5)+facet_wrap(~final_data$Gender)

```

 plot the association between quantity of smoking and developing lung cancer 

```{r warning=FALSE}

ggplot(data ,aes( final_data$label , final_data$Pack.Years.Smoked ,color = final_data$label)) +geom_boxplot(aes(color = final_data$label)) + mytheme+
geom_jitter(aes(fill = final_data$label))+ xlab("group") + ylab("smoking state")+
stat_summary(fun.y="mean", color="black", geom="point",shape=18, size=5,position=position_dodge(width=0.75))+
stat_compare_means( method = "t.test",size = 5)
```

 no association concluded between Pack Years Smoked and the outcome



